### DISCUSSION TITLE
MLMG solver failed to converge

### DISCUSSION THREAD

**Discussion by solarmaterial** [AMReX-Codes/amrex#4269] [Q&A]
Hi amrex team, 

I am using the MLMG linear solver (MLABecLaplacian) to solve a discretized heat conduction equation in a 2D domain in single level (uniform grid) with 2400 grids in the x-direction and 200 grids in the y-direction. An external heat source `q` is applied along the x-direction (`q=q(x,t`). When the magnitude of the heat source is relatively small, the solver performs well and converges quickly. However, when I increase the magnitude of the heat source by two orders of magnitude, the solver fails to achieve the required precision even after thousands of iterations.

I have tried several approaches to address this issue, including reducing the precision requirements, increasing the maximum number of iterations, and switching solvers (e.g., BiCGSTAB, smoother, and Hypre with preconditioners). Unfortunately, as the simulation progresses in time, the solver eventually fails because it cannot meet the required precision.

Interestingly, when I reduce the number of grids in the y-direction to 5 or 50 (while keeping the x-direction resolution unchanged), the solver is able to proceed, albeit requiring multiple iterations to converge. However, when the number of grids in the y-direction increases to 100 or 200, the solver fails.

Could you please provide suggestions or insights on how to address this issue? I greatly appreciate your help and look forward to your reply.

Yours sincerely,
solar

PS：The result is shown in the figure below.

![1734276028855](https://github.com/user-attachments/assets/8ba62103-1bd3-4ec4-90d0-13bd91ba8ed2)

**Comment by WeiqunZhang** [comment]
MLMG is a geometric multigrid method. It's most effective when (1) the cell size is uniform, (2) the domain is a cube, and (3) the number of cells in each direction is power of 2. What is the cell size in your problem? If you can, could you try to adjust your problem settings toward the three conditions above.

When you use hypre, did you set max_coarsening_level? If not, you should try to set LPInfo's max coarsening level to 0 when using hypre so that the entire problem is solved by hypre.

You could also try GMRES. See amrex/Tests/LinearSolvers/ABecLaplacian_C/ for an example.

**Reply by solarmaterial** [reply]
Thank you very much for your response. I will continue to explore based on your suggestions. Thanks again!

**Reply by solarmaterial** [reply]
Hello, Weiqun. After adjusting the grid numbers in the x and y directions to `nx cell = 2048` and `ny cell = 128` respectively (`both powers of 2`), I ran the calculation on a single node with `28` CPUs. I set `max_grid_size_x = 147` and `max_grid_size_y = 64`, which distributed the processors evenly across the computational domain in a layout of 14 tiles per row and 2 rows. Under these conditions, MLMG performed very smoothly, and the problem was resolved.
The result printed is something like
![image](https://github.com/user-attachments/assets/14a2f336-28a3-4544-a136-d793ee94250d)

However, when I switched to two nodes (`56 `CPUs), doubled the number of grids in the y direction to `ny cell = 256` (while keeping the x direction at `2048`), and used a tiling setup `with max_grid_size_x = 147` and `max_grid_size_y = 128`, the same issue reappeared.

The abnormal results are similar to those shown in the figure below.
![image](https://github.com/user-attachments/assets/13f224a0-67ca-47f9-9a7d-2bcdb4cd81b5)

Based on the outputs from both the normal and abnormal calculations, it seems that the number of `Final: Iteration` within the solver was `45` in one case and only `2` in the other. I suspect this might be the key factor affecting the solver's performance. Could this be due to a mismatch in grid sizes between the `x` and `y` directions caused by refining only the `y`-direction grid? Do you have any further suggestions on how to address this issue? Thank you again for your valuable guidance!

**Reply by WeiqunZhang** [reply]
max_grid_size of 147 is not an ideal number for multigrid. In fact, MLMG couldn't coarsen at all in this case and bicgstab was used for the entire problem. Please try something like 128, or 64 or 32.

Edit. I meant 147 is *not*.

**Reply by ajnonaka** [reply]
max_grid_size_x=147 is not a good choice - you want that to be a power of 2, or something where the prime factorization is as close to having only 2's in it as possible.  Otherwise the multigrid won't coarsen much and then you will require either a very large number of V-cycles or the bottom solver performance will suffer.  Try max_grid_size=128 (especially if n_cell_x=2048) then you will have 16 grids in the x-direction.

**Reply by solarmaterial** [reply]
Thank you for your help!! I will try

**Reply by solarmaterial** [reply]
When I set `max_grid_size_x=128` and `max_grid_size_y=64`, the solver failed. 
However, when I set `max_grid_size_x=147` and `max_grid_size_y=64`, the solver worked smoothly, which is very confusing.
I am using a single node with  28 cpus, and `ncell_x=2048`, `ncell_y=128`. Could this be related to the solver's DistributionMapping?

**Reply by WeiqunZhang** [reply]
> *It's most effective when (1) the cell size is uniform*

> *What is the cell size in your problem?*

**Reply by solarmaterial** [reply]
In this case (i.e., `ncell_x=2048  ncell_y=128`)，the cell size in the x-dir is `7.89e-6 `while the y-dir `7.81e-6`.

**Reply by WeiqunZhang** [reply]
What about the conductivity? Is it constant or a function of x and y? Have you tried GMRES?

Why 147 works is because it is actually not mulitgrid in that case. It's just bicgstab, which could work for small problems.

**Reply by solarmaterial** [reply]
The thermal conductivity coefficient is nonlinear, i.e., 𝑘=𝑘(𝑇), where 𝑇 is the temperature, which is the solution result of the previous time step. I use Newton iteration to update this coefficient, each iteration I am puzzled that 147 works, but 128 does not. I haven't tried GMRES yet, I will give it a try

**Reply by solarmaterial** [reply]
<img width="547" alt="88f2dc57ffe0587705d20e670558731" src="https://github.com/user-attachments/assets/ac3c18f2-9f3f-4c65-b068-0c885af99d14" />
It seems that GMRES is about to fail...

`nx_cell = 2048 ny_cell = 128 max_grid_size_x = 512 max_grid_size_y = 64`

**Reply by WeiqunZhang** [reply]
Sorry for not being clear in my previous messages. Stating things clearly was never my strength. We could try the following. (1) You provide a test code and I will try to investigate. (2) Or I can try again.

> When you use hypre, did you set max_coarsening_level? If not, you should try to set LPInfo's max coarsening level to 0 when using hypre so that the entire problem is solved by hypre.

Could you try the suggestion? (And forget about GMRES.) If you need more explanation, let me know. 

> You could also try GMRES. See amrex/Tests/LinearSolvers/ABecLaplacian_C/ for an example.

By GMRES, I mean the setup in `void MyTest::solveABecLaplacianGMRES ()` in amrex/Tests/LinearSolvers/ABecLaplacian_C/MyTest.cpp.

**Reply by solarmaterial** [reply]
So, you mean that instead of using the GMRES bottom solver in hypre, I should use the GMRES solver in MLMG, right?Anyway, thank you so much for your help! :)  I will give it a try. Also the max_coarsening_level =  0.
I will compile a test code later. Thanks again for your suggestions!!

**Reply by solarmaterial** [reply]
Hi, Weiqun
The code link is here https://github.com/solarmaterial/amrex-heat-equation).

**Reply by WeiqunZhang** [reply]
The code you shared above contains a single function without any other functions (including main) or input data. I am not sure how I could help you. By "providing a test code", I meant a complete code that I could `make` and then run.

```
void heatConduction (MultiFab& phi_old,
              MultiFab& phi_new,
              MultiFab& Laser_rhs,
              MultiFab& temp_new,
              Real dt,
              const Geometry& geom,
              const BoxArray& grids,
              const DistributionMapping& dmap,
              const Vector<BCRec>& bc,
              Real time,
              int ny_cell,
              //amrex::Vector<int>& HeatSource,
              amrex::Vector<amrex::Real>& First_phy_HeatSource)
{ ... }
```

**Reply by solarmaterial** [reply]
Hi, Weiqun
The complete code was uploaded.
The code link is here https://github.com/solarmaterial/amrex-heat-equation).

**Reply by solarmaterial** [reply]
Hi, Weiqun
Could you please let me know if you’ve identified any issues in the code I provided? I’d greatly appreciate any insights or advice you could share. Unfortunately, I’m still struggling to resolve the problem.

Thank you for your time and guidance!
